spec/large_clusters_spec.rb

```diff
@@ -0,0 +1,90 @@
+# frozen_string_literal: true
+
+require "cluster"
+require "cluster_ht_item"
+require "large_clusters"
+require "large_cluster_error"
+require "spec_helper"
+
+RSpec.describe LargeClusters do
+  let(:mock_data) { [1_759_445, 8_878_489].to_set }
+  let(:large_clusters) { described_class.new(mock_data) }
+
+  before(:each) do
+    Cluster.each(&:delete)
+  end
+
+  describe "#ocns" do
+    it "has the list of ocns provided" do
+      expect(large_clusters.ocns).to include(1_759_445)
+    end
+  end
+
+  describe "#load_large_clusters" do
+    it "pulls the list of ocns from the ENV file" do
+      ENV["LARGE_CLUSTER_OCNS"] = "/tmp/large_cluster_ocns.txt"
+      `echo "1001117803" > /tmp/large_cluster_ocns.txt`
+      large_clusters = described_class.new
+      expect(large_clusters.ocns).to include(1_001_117_803)
+    end
+  end
+
+  describe "reclustering of large clusters" do
+    # A change in the OCNs of ht_items' OCNs or ocn_resolutions can result in a recluster.
+    # Reclustering a "large cluster" will result in a cluster with incomplete holdings.
+    let(:ht1) { build(:ht_item, ocns: [large_clusters.ocns.first]) }
+    let(:ht2) { build(:ht_item, ocns: [large_clusters.ocns.first]) }
+    let(:ht3) { build(:ht_item, ocns: [1]) }
+
+    it "does not raise an error when performing a simple update" do
+      ClusterHtItem.new(ht1).cluster
+      ClusterHtItem.new(ht2).cluster
+      ht2.ocns << 1
+      expect { ClusterHtItem.new(ht2).cluster }.not_to raise_error
+    end
+
+    it "raises a LargeClusterError when reclustering a large cluster" do
+      ClusterHtItem.new(ht1).cluster
+      ClusterHtItem.new(ht2).cluster
+      expect(Cluster.first.ht_items.count).to eq(2)
+      ht2.ocns = [1]
+      expect do
+        ClusterHtItem.new(ht2).cluster
+      end.to raise_exception(LargeClusterError)
+    end
+  end
+
+  describe "merging of large clusters" do
+    # Merging a non-"large cluster" into a "large cluster" means future holdings will be deduped.
```

Comment:
I think this behavior is implied from the fact that we recluster by reloading constituent data and the update spec for deduping above. I think it's probably worth a spec to verify this behavior given the desire to have this comment and that it shouldn't need much additional setup beyond what's here -- i.e. create a non-large cluster, give it two holdings that should be collapsed, merge it into an empty large cluster, verify the holdings were collapsed.

Reply:
As far as I can tell, a cluster merge doesn't perform a recluster. It performs `add_members_from` which is `push_to_field`. Holdings won't be de-duped on merge, only future holdings.

Reply:
hmm, ok. Not sure that's the behavior we want. Let's discuss Friday.

Reply:
Discussed and we are OK with some potential duplicates since it doesn't affect reports and simplifies implementation.